akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala

```diff
@@ -658,26 +662,29 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
   final case class IncomingConnection(
     localAddress: InetSocketAddress,
     remoteAddress: InetSocketAddress,
-    flow: Flow[HttpResponse, HttpRequest, Unit]) {
+    flow: Flow[HttpResponse, HttpRequest, Promise[Option[Option[InetSocketAddress]]]]) {
 
     /**
      * Handles the connection with the given flow, which is materialized exactly once
      * and the respective materialization result returned.
      */
-    def handleWith[Mat](handler: Flow[HttpRequest, HttpResponse, Mat])(implicit fm: Materializer): Mat =
-      flow.joinMat(handler)(Keep.right).run()
+    def handleWith[Mat](handler: Flow[HttpRequest, HttpResponse, Mat])(implicit fm: Materializer): Mat = {
+      val fillMeInAndUserMat = flow.joinMat(handler)(Keep.both).run()
+      fillMeInAndUserMat._1.success(Some(Some(remoteAddress)))
+      fillMeInAndUserMat._2
+    }
```

Comment:
> Absolutely. But using the Attribute means you first of all have to know that it exists and that you may want to set it. Right now it becomes a hidden setting. Do you agree with that assessment?

Right, that is certainly true: someone needs to set it, correct. We need to do this for each materialization (connection).
It is different in semantics though (and I do think we prefer the attribute-way semantics):

- attribute when I'd forget to set, we're in the (pseudocode) `None[Option[InetSocketAddress]], and I can happily continue without it.
  - forgetting to complete the materialized promise causes the server to lock-up waiting for it. It's tricky to "oh by default we'll emit None", since the completion must happen after materialization, so there must be a delay by construction.

A number of other downsides of the element injecting were mentioned in this PR as well (inherent measurable delay, increasing the burden of the server's type signature) and explained the both approaches to both @rkuhn, @drewkh and @jrudolph which also think this is the lesser evil. 

So I hope your argument of picking the lesser of two evils is satisfied and we can move on - there's loads of more work here, and likely more tricky than this simple one :-) I got the WS TCK running on jenkins, so feel more comfortable to tackle part 2 of this effort now, which relates to making the user provided WS accessible in a reusable blueprint – _that_ will be way more tricky than this was I think.